test/base/src/androidTest/kotlin/io/realm/test/shared/RealmTests.kt

```diff
@@ -402,6 +406,93 @@ class RealmTests {
             }
     }
 
+    @Test
+    @Suppress("LongMethod")
+    fun deleteRealm() {
+        val fileSystem = FileSystem.SYSTEM
+        val testDir = PlatformUtils.createTempDir("test_dir")
+        val testDirPath = testDir.toPath()
+        assertTrue(fileSystem.exists(testDirPath))
+
+        val configuration = RealmConfiguration.Builder(schema = setOf(Parent::class, Child::class))
+            .path("$testDir/default.realm")
+            .build()
+
+        val bgThreadReadyChannel = Channel<Unit>(1)
+        val readyToCloseChannel = Channel<Unit>(1)
+        val closedChannel = Channel<Unit>(1)
+
+        runBlocking {
+            val testRealm = Realm.open(configuration)
+
+            val deferred = async {
+                // Create another Realm to ensure the log files are generated.
+                val anotherRealm = Realm.open(configuration)
+                bgThreadReadyChannel.send(Unit)
+
+                readyToCloseChannel.receive()
+
+                anotherRealm.close()
+                closedChannel.send(Unit)
+            }
+
+            // Waits for background thread opening the same Realm.
+            bgThreadReadyChannel.receive()
+
+            // Check the realm got created correctly and signal that it can be closed.
+            fileSystem.list(testDirPath)
+                .also { testDirPathList ->
+                    assertEquals(4, testDirPathList.size) // db file, .lock, .management, .note
+                    readyToCloseChannel.send(Unit)
+                }
+
+            testRealm.close()
+
+            closedChannel.receive()
+
+            // Delete realm now that it's fully closed.
+            Realm.deleteRealm(configuration)
+
+            // Lock file should never be deleted.
+            fileSystem.list(testDirPath)
+                .also { testDirPathList ->
+                    assertEquals(1, testDirPathList.size) // only .lock file remains
+
+                    assertTrue(fileSystem.exists("${configuration.path}.lock".toPath()))
+                }
+
+            deferred.cancel()
+            bgThreadReadyChannel.close()
+            readyToCloseChannel.close()
+            closedChannel.close()
+        }
+    }
+
+    @Test
+    fun deleteRealm_failures() {
+        val tempDirA = PlatformUtils.createTempDir()
+
+        val configA = RealmConfiguration.Builder(schema = setOf(Parent::class, Child::class))
+            .path("$tempDirA/anotherRealm.realm")
+            .build()
+
+        // Creates a new Realm file.
+        val anotherRealm = Realm.open(configA)
+
+        // Deleting it without having closed it should fail.
+        assertFailsWith<IllegalStateException> {
```

Comment:
Maybe assert on the actual message too with `assertFailsWith` because the implementation seems to differ on the two platforms and would actually yield different exceptions.